File: avalon-sandbox/avalon-net/Composition/ModelDefault/DefaultContainmentModelAssemblyHelper.cs
/* 
* Copyright 2003-2004 The Apache Software Foundation
* Licensed  under the  Apache License,  Version 2.0  (the "License");
* you may not use  this file  except in  compliance with the License.
* You may obtain a copy of the License at 
* 
*   http://www.apache.org/licenses/LICENSE-2.0
* 
* Unless required by applicable law or agreed to in writing, software
* distributed  under the  License is distributed on an "AS IS" BASIS,
* WITHOUT  WARRANTIES OR CONDITIONS  OF ANY KIND, either  express  or
* implied.
* 
* See the License for the specific language governing permissions and
* limitations under the License.
*/

namespace Apache.Avalon.Composition.Model.Default
{
	using System;
	
	using Apache.Avalon.Meta;
	using Apache.Avalon.Composition.Data;
	
	/// <summary> A utility class that assists in the location of a model relative
	/// a supplied path.
	/// 
	/// </summary>
	/// <author>  <a href="mailto:dev@avalon.apache.org">Avalon Development Team</a>
	/// </author>
	/// <version>  $Revision: 1.1 $ $Date: 2004/02/28 22:15:42 $
	/// </version>
	class DefaultContainmentModelAssemblyHelper
	{
		//-------------------------------------------------------------------
		// static
		//-------------------------------------------------------------------
		
		//UPGRADE_NOTE: Final was removed from the declaration of 'REZ '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
		//UPGRADE_NOTE: The initialization of  'REZ' was moved to static method 'Apache.Avalon.Composition.Model.Default.DefaultContainmentModelAssemblyHelper'. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1005"'
		// private static readonly Resources REZ;
		
		//-------------------------------------------------------------------
		// immutable state
		//-------------------------------------------------------------------
		
		//UPGRADE_NOTE: Final was removed from the declaration of 'm_context '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
		private IContainmentContext m_context;
		//UPGRADE_NOTE: Final was removed from the declaration of 'm_model '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
		private DefaultContainmentModel m_model;
		
		//-------------------------------------------------------------------
		// constructor
		//-------------------------------------------------------------------
		
		public DefaultContainmentModelAssemblyHelper(IContainmentContext context, DefaultContainmentModel model)
		{
			m_context = context;
			m_model = model;
		}
		
		//-------------------------------------------------------------------
		// implementation
		//-------------------------------------------------------------------
		
		/// <summary> Assemble a target model during which all deployment and runtime 
		/// dependencies are assigned a provider model.
		/// 
		/// </summary>
		/// <param name="model">the target model to be assembled
		/// </param>
		/// <param name="subject">the model requesting the assembly
		/// </param>
		//UPGRADE_ISSUE: Class hierarchy differences between ''java.util.List'' and ''System.Collections.IList'' may cause compilation errors. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1186"'
		public virtual void  assembleModel(IDeploymentModel model, System.Collections.IList subjects)
		{
			if (null == model)
			{
				throw new System.ArgumentNullException("model");
			}
			if (null == subjects)
			{
				throw new System.ArgumentNullException("subjects");
			}
			if (subjects.Contains(model))
			{
				return ;
			}
			if (model.IsAssembled)
			{
				return ;
			}
			
			if (model is IComponentModel)
			{
				AssembleComponent((IComponentModel) model, subjects);
			}
			else
			{
				model.Assemble(subjects);
			}
		}
		
		//UPGRADE_ISSUE: Class hierarchy differences between ''java.util.List'' and ''System.Collections.IList'' may cause compilation errors. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1186"'
		private void AssembleComponent(IComponentModel model, System.Collections.IList subjects)
		{
			IModelRepository repository = m_context.ModelRepository;
			
			//
			// locate and assemble the component context handler
			//
			
			if (model.ContextModel != null)
			{
				IContextModel context = model.ContextModel;
				System.Type type = context.StrategyClass;
				if (!type.FullName.Equals(Apache.Avalon.Composition.Model.IContextModel_Fields.DEFAULT_STRATEGY_CLASSNAME))
				{
					if (null == context.Provider)
					{
						try
						{
							//UPGRADE_TODO: The equivalent in .NET for method 'java.util.List.add' may return a different value. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1043"'
							subjects.Add(model);
							StageDescriptor stage = new StageDescriptor(type.FullName);
							IDeploymentModel provider = findExtensionProvider(repository, stage, subjects);
							context.Provider = provider;
						}
						//UPGRADE_NOTE: Exception 'java.lang.Throwable' was converted to 'System.Exception' which has different behavior. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1100"'
						catch (System.Exception e)
						{
							//UPGRADE_NOTE: Final was removed from the declaration of 'error '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
							//UPGRADE_TODO: The equivalent in .NET for method 'java.lang.Object.toString' may return a different value. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1043"'
							String error = "Unable to assemble component: " + model + " due to a component context phase handler establishment failure.";
							throw new AssemblyException(error, e);
						}
						finally
						{
							//UPGRADE_TODO: Method 'java.util.List.remove' was converted to 'System.Collections.IList.Remove' which has a different behavior. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1073_javautilListremove_javalangObject"'
							subjects.Remove(model);
						}
					}
				}
			}
			
			//
			// locate and resolve the stage providers
			//
			
			IStageModel[] stages = model.StageModels;
			for (int i = 0; i < stages.Length; i++)
			{
				IStageModel stage = stages[i];
				if (null == stage.Provider)
				{
					try
					{
						//UPGRADE_TODO: The equivalent in .NET for method 'java.util.List.add' may return a different value. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1043"'
						subjects.Add(model);
						IDeploymentModel provider = findExtensionProvider(repository, stage, subjects);
						stage.Provider = provider;
					}
					//UPGRADE_NOTE: Exception 'java.lang.Throwable' was converted to 'System.Exception' which has different behavior. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1100"'
					catch (System.Exception e)
					{
						//UPGRADE_NOTE: Final was removed from the declaration of 'error '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
						//UPGRADE_TODO: The equivalent in .NET for method 'java.lang.Object.toString' may return a different value. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1043"'
						String error = "Unable to assemble component: " + model + " due to a component extension handler establishment failure.";
						throw new AssemblyException(error, e);
					}
					finally
					{
						//UPGRADE_TODO: Method 'java.util.List.remove' was converted to 'System.Collections.IList.Remove' which has a different behavior. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1073_javautilListremove_javalangObject"'
						subjects.Remove(model);
					}
				}
			}
			
			//
			// locate and resolve the service providers
			//
			
			IDependencyModel[] dependencies = model.DependencyModels;
			for (int i = 0; i < dependencies.Length; i++)
			{
				IDependencyModel dependency = dependencies[i];
				if (null == dependency.Provider)
				{
					try
					{
						//UPGRADE_TODO: The equivalent in .NET for method 'java.util.List.add' may return a different value. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1043"'
						subjects.Add(model);
						IDeploymentModel provider = findDependencyProvider(repository, dependency, subjects);
						dependency.Provider = provider;
					}
					//UPGRADE_NOTE: Exception 'java.lang.Throwable' was converted to 'System.Exception' which has different behavior. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1100"'
					catch (System.Exception e)
					{
						//UPGRADE_NOTE: Final was removed from the declaration of 'error '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
						//UPGRADE_TODO: The equivalent in .NET for method 'java.lang.Object.toString' may return a different value. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1043"'
						String error = "Unable to assemble component: " + model + " due to a service provider establishment failure.";
						throw new AssemblyException(error, e);
					}
					finally
					{
						//UPGRADE_TODO: Method 'java.util.List.remove' was converted to 'System.Collections.IList.Remove' which has a different behavior. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1073_javautilListremove_javalangObject"'
						subjects.Remove(model);
					}
				}
			}
		}
		
		//UPGRADE_ISSUE: Class hierarchy differences between ''java.util.List'' and ''System.Collections.IList'' may cause compilation errors. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1186"'
		private IDeploymentModel findDependencyProvider(IModelRepository repository, IDependencyModel dependency, System.Collections.IList subjects)
		{
			String path = dependency.Path;
			if (null != (System.Object) path)
			{
				IDeploymentModel model = m_model.GetModel(path);
				if (null == model)
				{
					//UPGRADE_NOTE: Final was removed from the declaration of 'error '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
					//UPGRADE_TODO: The equivalent in .NET for method 'java.lang.Object.toString' may return a different value. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1043"'
					String error = "Could not locate a model at the address: [" + path + "] in " + this + ".";
					throw new AssemblyException(error);
				}
				assembleModel(model, subjects);
				return model;
			}
			else
			{
				return findDependencyProvider(repository, dependency.Dependency, subjects);
			}
		}
		
		//UPGRADE_ISSUE: Class hierarchy differences between ''java.util.List'' and ''System.Collections.IList'' may cause compilation errors. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1186"'
		private IDeploymentModel findDependencyProvider(IModelRepository repository, DependencyDescriptor dependency, System.Collections.IList subjects)
		{
			IDeploymentModel[] candidates = repository.GetCandidateProviders(dependency);
			IModelSelector selector = new DefaultModelSelector();
			IDeploymentModel model = selector.Select(candidates, dependency);
			if (model != null)
			{
				assembleModel(model, subjects);
				return model;
			}
			
			//
			// otherwise, check for any packaged profiles that 
			// we could use to construct the model
			//
			
			DeploymentProfile[] profiles = findDependencyProfiles(dependency);
			IProfileSelector profileSelector = new DefaultProfileSelector();
			DeploymentProfile profile = profileSelector.Select(profiles, dependency);
			if (profile != null)
			{
				try
				{
					IDeploymentModel solution = m_model.CreateDeploymentModel(profile);
					assembleModel(solution, subjects);
					m_model.AddModel(solution);
					return solution;
				}
				catch (AssemblyException ae)
				{
					//UPGRADE_NOTE: Final was removed from the declaration of 'error '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
					//UPGRADE_TODO: The equivalent in .NET for method 'java.lang.Object.toString' may return a different value. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1043"'
					String error = "Nested assembly failure while attempting to construct model" + " for the profile: [" + profile + "] for the dependency: [" + dependency + "].";
					throw new AssemblyException(error, ae);
				}
				catch (ModelException me)
				{
					//UPGRADE_NOTE: Final was removed from the declaration of 'error '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
					//UPGRADE_TODO: The equivalent in .NET for method 'java.lang.Object.toString' may return a different value. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1043"'
					String error = "Nested model failure while attempting to add model" + " for the profile: [" + profile + "] for the dependency: [" + dependency + "].";
					throw new AssemblyException(error, me);
				}
			}
			else
			{
				//UPGRADE_NOTE: Final was removed from the declaration of 'error '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
				String error = "Unable to locate a service provider for the dependency: [ " + dependency + "].";
				throw new AssemblyException(error);
			}
		}
		
		//UPGRADE_ISSUE: Class hierarchy differences between ''java.util.List'' and ''System.Collections.IList'' may cause compilation errors. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1186"'
		private IDeploymentModel findExtensionProvider(IModelRepository repository, IStageModel stage, System.Collections.IList subjects)
		{
			String path = stage.Path;
			if (null != (System.Object) path)
			{
				IDeploymentModel model = m_model.GetModel(path);
				if (null == model)
				{
					//UPGRADE_NOTE: Final was removed from the declaration of 'error '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
					//UPGRADE_TODO: The equivalent in .NET for method 'java.lang.Object.toString' may return a different value. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1043"'
					String error = "Could not locate a model at the address: [" + path + "] in " + this + ".";
					throw new AssemblyException(error);
				}
				assembleModel(model, subjects);
				return model;
			}
			else
			{
				return findExtensionProvider(repository, stage.Stage, subjects);
			}
		}
		
		//UPGRADE_ISSUE: Class hierarchy differences between ''java.util.List'' and ''System.Collections.IList'' may cause compilation errors. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1186"'
		private IDeploymentModel findExtensionProvider(IModelRepository repository, StageDescriptor stage, System.Collections.IList subjects)
		{
			IDeploymentModel[] candidates = repository.GetCandidateProviders(stage);
			IModelSelector selector = new DefaultModelSelector();
			IDeploymentModel model = selector.Select(candidates, stage);
			if (model != null)
			{
				assembleModel(model, subjects);
				return model;
			}
			
			//
			// otherwise, check for any packaged profiles that 
			// we could use to construct the model
			//
			
			DeploymentProfile[] profiles = findExtensionProfiles(stage);
			IProfileSelector profileSelector = new DefaultProfileSelector();
			DeploymentProfile profile = profileSelector.Select(profiles, stage);
			if (profile != null)
			{
				try
				{
					IDeploymentModel solution = m_model.CreateDeploymentModel(profile);
					assembleModel(solution, subjects);
					m_model.AddModel(solution);
					return solution;
				}
				catch (AssemblyException ae)
				{
					//UPGRADE_NOTE: Final was removed from the declaration of 'error '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
					//UPGRADE_TODO: The equivalent in .NET for method 'java.lang.Object.toString' may return a different value. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1043"'
					String error = "Nested assembly failure while attempting to construct model" + " for the extension profile: [" + profile + "] for the stage dependency: [" + stage + "].";
					throw new AssemblyException(error, ae);
				}
				catch (ModelException me)
				{
					//UPGRADE_NOTE: Final was removed from the declaration of 'error '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
					//UPGRADE_TODO: The equivalent in .NET for method 'java.lang.Object.toString' may return a different value. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1043"'
					String error = "Nested model failure while attempting to add model" + " for the extension profile: [" + profile + "] for the stage dependency: [" + stage + "].";
					throw new AssemblyException(error, me);
				}
			}
			else
			{
				//UPGRADE_NOTE: Final was removed from the declaration of 'error '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
				String error = "Unable to locate a extension provider for the stage: [ " + stage + "].";
				throw new AssemblyException(error);
			}
		}
		
		private DeploymentProfile[] findExtensionProfiles(StageDescriptor stage)
		{
			ITypeRepository repository = m_context.TypeLoaderModel.TypeRepository;
			TypeDescriptor[] types = repository.GetTypes(stage);
			try
			{
				return getProfiles(repository, types);
			}
			catch (TypeUnknownException tue)
			{
				// will not happen
				//UPGRADE_NOTE: Final was removed from the declaration of 'error '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
				String error = "An irrational condition has occured.";
				throw new ModelRuntimeException(error, tue);
			}
		}
		
		private DeploymentProfile[] findDependencyProfiles(DependencyDescriptor dependency)
		{
			ITypeRepository repository = m_context.TypeLoaderModel.TypeRepository;
			TypeDescriptor[] types = repository.GetTypes(dependency);
			try
			{
				return getProfiles(repository, types);
			}
			catch (TypeUnknownException tue)
			{
				// will not happen
				//UPGRADE_NOTE: Final was removed from the declaration of 'error '. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1003"'
				String error = "An irrational condition has occured.";
				throw new ModelRuntimeException(error, tue);
			}
		}
		
		private DeploymentProfile[] getProfiles(ITypeRepository repository, TypeDescriptor[] types)
		{
			//UPGRADE_ISSUE: Class hierarchy differences between 'java.util.ArrayList' and 'System.Collections.ArrayList' may cause compilation errors. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1186"'
			System.Collections.ArrayList list = new System.Collections.ArrayList();
			for (int i = 0; i < types.Length; i++)
			{
				DeploymentProfile[] profiles = repository.GetProfiles(types[i]);
				for (int j = 0; j < profiles.Length; j++)
				{
					//UPGRADE_TODO: The equivalent in .NET for method 'java.util.ArrayList.add' may return a different value. 'ms-help://MS.VSCC.2003/commoner/redir/redirect.htm?keyword="jlca1043"'
					list.Add(profiles[j]);
				}
			}
			return (DeploymentProfile[]) list.ToArray( typeof(DeploymentProfile) );
		}
	}
}

File: avalon/composition/impl/src/java/org/apache/avalon/composition/model/impl/DefaultContainmentModelAssemblyHelper.java
/* 
 * Copyright 2004 Apache Software Foundation
 * Licensed  under the  Apache License,  Version 2.0  (the "License");
 * you may not use  this file  except in  compliance with the License.
 * You may obtain a copy of the License at 
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed  under the  License is distributed on an "AS IS" BASIS,
 * WITHOUT  WARRANTIES OR CONDITIONS  OF ANY KIND, either  express  or
 * implied.
 * 
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.avalon.composition.model.impl;

import java.util.ArrayList;

import org.apache.avalon.composition.data.DeploymentProfile;
import org.apache.avalon.composition.model.ContainmentModel;
import org.apache.avalon.composition.model.ComponentModel;
import org.apache.avalon.composition.model.DeploymentModel;
import org.apache.avalon.composition.model.AssemblyException;
import org.apache.avalon.composition.model.ModelRepository;
import org.apache.avalon.composition.model.ContextModel;
import org.apache.avalon.composition.model.DependencyModel;
import org.apache.avalon.composition.model.StageModel;
import org.apache.avalon.composition.model.TypeUnknownException;
import org.apache.avalon.composition.model.ModelRuntimeException;
import org.apache.avalon.composition.model.ModelException;
import org.apache.avalon.composition.model.TypeRepository;
import org.apache.avalon.composition.model.ModelSelector;
import org.apache.avalon.composition.model.ProfileSelector;
import org.apache.avalon.composition.provider.ContainmentContext;

import org.apache.avalon.excalibur.i18n.ResourceManager;
import org.apache.avalon.excalibur.i18n.Resources;

import org.apache.avalon.meta.info.Type;
import org.apache.avalon.meta.info.DependencyDescriptor;
import org.apache.avalon.meta.info.ServiceDescriptor;
import org.apache.avalon.meta.info.StageDescriptor;

/**
 * A utility class that assists in the location of a model relative
 * a supplied path.
 *
 * @author <a href="mailto:dev@avalon.apache.org">Avalon Development Team</a>
 * @version $Revision: 1.2 $ $Date: 2004/02/10 16:23:33 $
 */
class DefaultContainmentModelAssemblyHelper
{
    //-------------------------------------------------------------------
    // static
    //-------------------------------------------------------------------

    private static final Resources REZ =
      ResourceManager.getPackageResources( 
        DefaultContainmentModelAssemblyHelper.class );

    //-------------------------------------------------------------------
    // immutable state
    //-------------------------------------------------------------------

    private final ContainmentContext m_context;
    private final ContainmentModel m_model;

    //-------------------------------------------------------------------
    // constructor
    //-------------------------------------------------------------------

    public DefaultContainmentModelAssemblyHelper( 
      ContainmentContext context, ContainmentModel model )
    {
        m_context = context;
        m_model = model;
    }

    //-------------------------------------------------------------------
    // implementation
    //-------------------------------------------------------------------

    public void assembleModel( DeploymentModel model ) 
      throws AssemblyException
    {
         if( null == model )
         {
             throw new NullPointerException( "model" );
         }

         if( model.isAssembled() ) 
         {
             return;
         }
         else
         {
             if( model instanceof ComponentModel )
             {
                 assembleComponent( (ComponentModel) model );
             }
             else
             {
                 model.assemble();
             }
        }
    }

    private void assembleComponent( ComponentModel model ) throws AssemblyException
    {
        ModelRepository repository = m_context.getModelRepository();

        //
        // locate and assemble the component context handler
        //

        if( model.getContextModel() != null )
        {
            ContextModel context = model.getContextModel();
            Class clazz = context.getStrategyClass();
            if( !clazz.getName().equals( 
              ContextModel.DEFAULT_STRATEGY_CLASSNAME ) )
            {
                if( null == context.getProvider() )
                {
                    try
                    {
                        StageDescriptor stage = 
                          new StageDescriptor( clazz.getName() );
                        DeploymentModel provider = 
                          findExtensionProvider( repository, stage );
                        context.setProvider( provider );
                    }
                    catch( Throwable e )
                    {
                        final String error = 
                          "Unable to assemble component: " 
                          + model 
                         + " due to a component context phase handler establishment failure.";
                        throw new AssemblyException( error, e );
                    }
                }
            }
        }

        //
        // locate and resolve the stage providers
        //

        StageModel[] stages = model.getStageModels();
        for( int i=0; i<stages.length; i++ )
        {
            StageModel stage = stages[i];
            if( null == stage.getProvider() )
            {
                try
                {
                    DeploymentModel provider =
                      findExtensionProvider( repository, stage );
                    stage.setProvider( provider );
                }
                catch( Throwable e )
                {
                    final String error = 
                      "Unable to assemble component: " 
                      + model 
                      + " due to a component extension handler establishment failure.";
                    throw new AssemblyException( error, e );
                }
            }
        }

        //
        // locate and resolve the service providers
        //

        DependencyModel[] dependencies = model.getDependencyModels();
        for( int i=0; i<dependencies.length; i++ )
        {
            DependencyModel dependency = dependencies[i];
            if( null == dependency.getProvider() )
            {
                try
                {
                    DeploymentModel provider =
                      findDependencyProvider( repository, dependency );
                    dependency.setProvider( provider );
                }
                catch( Throwable e )
                {
                    final String error = 
                      "Unable to assemble component: " + model 
                      + " due to a service provider establishment failure.";
                    throw new AssemblyException( error, e );
                }
            }
        }
    }

    private DeploymentModel findDependencyProvider( 
      ModelRepository repository, DependencyModel dependency )
      throws AssemblyException
    {
        String path = dependency.getPath();
        if( null != path )
        {
            DeploymentModel model = m_model.getModel( path );
            if( null == model )
            {
                final String error = 
                  "Could not locate a model at the address: [" 
                  + path + "] in " + this + ".";
                throw new AssemblyException( error );
            }
            assembleModel( model );
            return model;
        }
        else
        {
            return findDependencyProvider( 
              repository, dependency.getDependency() );
        }
    }

    private DeploymentModel findDependencyProvider( 
      ModelRepository repository, DependencyDescriptor dependency )
      throws AssemblyException
    {
        DeploymentModel[] candidates = 
          repository.getCandidateProviders( dependency );
        ModelSelector selector = new DefaultModelSelector();
        DeploymentModel model = selector.select( candidates, dependency );
        if( model != null )
        {
            assembleModel( model );
            return model;
        }

        //
        // otherwise, check for any packaged profiles that 
        // we could use to construct the model
        //

        DeploymentProfile[] profiles = findDependencyProfiles( dependency );
        ProfileSelector profileSelector = new DefaultProfileSelector();
        DeploymentProfile profile = profileSelector.select( profiles, dependency );
        if( profile != null ) 
        {
            try
            {
                DeploymentModel solution = m_model.addModel( profile );
                assembleModel( solution );
                return solution;
            }
            catch( AssemblyException ae )
            {
                final String error = 
                  "Nested assembly failure while attempting to construct model"
                  + " for the profile: [" + profile + "] for the dependency: ["
                  + dependency + "].";
                throw new AssemblyException( error, ae );
            }
            catch( ModelException me )
            {
                final String error = 
                  "Nested model failure while attempting to add model"
                  + " for the profile: [" + profile + "] for the dependency: ["
                  + dependency + "].";
                throw new AssemblyException( error, me );
            }
        }
        else
        {
            final String error = 
              "Unable to locate a service provider for the dependency: [ "
              + dependency + "].";
            throw new AssemblyException( error );
        }
    }

    private DeploymentModel findExtensionProvider( 
      ModelRepository repository, StageModel stage )
      throws AssemblyException
    {
        String path = stage.getPath();
        if( null != path )
        {
            DeploymentModel model = m_model.getModel( path );
            if( null == model )
            {
                final String error = 
                  "Could not locate a model at the address: [" 
                  + path + "] in " + this + ".";
                throw new AssemblyException( error );
            }
            assembleModel( model );
            return model;
        }
        else
        {
            return findExtensionProvider( repository, stage.getStage() );
        }
    }

    private DeploymentModel findExtensionProvider( 
      ModelRepository repository, StageDescriptor stage )
      throws AssemblyException
    {
        DeploymentModel[] candidates = 
          repository.getCandidateProviders( stage );
        ModelSelector selector = new DefaultModelSelector();
        DeploymentModel model = selector.select( candidates, stage );
        if( model != null )
        {
            assembleModel( model );
            return model;
        }

        //
        // otherwise, check for any packaged profiles that 
        // we could use to construct the model
        //

        DeploymentProfile[] profiles = findExtensionProfiles( stage );
        ProfileSelector profileSelector = new DefaultProfileSelector();
        DeploymentProfile profile = profileSelector.select( profiles, stage );
        if( profile != null ) 
        {
            try
            {
                DeploymentModel solution = m_model.addModel( profile );
                assembleModel( solution );
                return solution;
            }
            catch( AssemblyException ae )
            {
                final String error = 
                  "Nested assembly failure while attempting to construct model"
                  + " for the extension profile: [" + profile 
                  + "] for the stage dependency: ["
                  + stage + "].";
                throw new AssemblyException( error, ae );
            }
            catch( ModelException me )
            {
                final String error = 
                  "Nested model failure while attempting to add model"
                  + " for the extension profile: [" + profile 
                  + "] for the stage dependency: ["
                  + stage + "].";
                throw new AssemblyException( error, me );
            }
        }
        else
        {
            final String error = 
              "Unable to locate a extension provider for the stage: [ "
              + stage + "].";
            throw new AssemblyException( error );
        }
    }

    private DeploymentProfile[] findExtensionProfiles( StageDescriptor stage )
    {
        TypeRepository repository = m_context.getClassLoaderModel().getTypeRepository();
        Type[] types = repository.getTypes( stage );
        try
        {
            return getProfiles( repository, types );
        }
        catch( TypeUnknownException tue )
        {
            // will not happen
            final String error = "An irrational condition has occured.";
            throw new ModelRuntimeException( error, tue );
        }
    }

    private DeploymentProfile[] findDependencyProfiles( DependencyDescriptor dependency )
    {
        TypeRepository repository = m_context.getClassLoaderModel().getTypeRepository();
        Type[] types = repository.getTypes( dependency );
        try
        {
            return getProfiles( repository, types );
        }
        catch( TypeUnknownException tue )
        {
            // will not happen
            final String error = "An irrational condition has occured.";
            throw new ModelRuntimeException( error, tue );
        }
    }

    private DeploymentProfile[] getProfiles( TypeRepository repository, Type[] types )
      throws TypeUnknownException
    {
        ArrayList list = new ArrayList();
        for( int i=0; i<types.length; i++ )
        {
            DeploymentProfile[] profiles = 
            repository.getProfiles( types[i] );
            for( int j=0; j<profiles.length; j++ )
            {
                list.add( profiles[j] );
            }
        }
        return (DeploymentProfile[]) list.toArray( new DeploymentProfile[0] );
    }

}
